File: cpp/src/geometry.cpp
#include "procvd/geometry.hpp"

#include <algorithm>

#include <opencv2/calib3d.hpp>
#include <opencv2/imgproc.hpp>

// ChArUco changed module and API in OpenCV 4.7. Support both so the project
// builds across versions:
//   4.7+  -> cv::aruco::CharucoDetector, CharucoBoard ctor, generateImage (objdetect)
//   <=4.6 -> CharucoBoard::create, detectMarkers + interpolateCornersCharuco (aruco)
#if (CV_VERSION_MAJOR > 4) || (CV_VERSION_MAJOR == 4 && CV_VERSION_MINOR >= 7)
#  define PROCVD_ARUCO_NEW 1
#  include <opencv2/objdetect.hpp>
#else
#  define PROCVD_ARUCO_NEW 0
#  include <opencv2/aruco.hpp>
#  include <opencv2/aruco/charuco.hpp>
#endif

#include "procvd/camera.hpp"
#include "procvd/projector.hpp"

namespace procvd {
namespace {

// Projector-space positions of the inner ChArUco corners, keyed by corner id
// (row-major, matching the ArUco convention). Pure — no hardware.
std::vector<cv::Point2f> proj_corner_table(const Config& cfg, int sq, int ox, int oy) {
    std::vector<cv::Point2f> table;
    table.reserve(static_cast<size_t>(cfg.board_cols) * cfg.board_rows);
    for (int r = 0; r < cfg.board_rows; ++r) {
        for (int c = 0; c < cfg.board_cols; ++c) {
            table.emplace_back(static_cast<float>(ox + (c + 1) * sq),
                               static_cast<float>(oy + (r + 1) * sq));
        }
    }
    return table;
}

}  // namespace

std::optional<Homography>
homography_from_correspondences(const std::vector<cv::Point2f>& proj_pts,
                                const std::vector<cv::Point2f>& cam_pts) {
    if (proj_pts.size() < 4 || proj_pts.size() != cam_pts.size()) {
        return std::nullopt;
    }
    const cv::Mat H = cv::findHomography(proj_pts, cam_pts, cv::RANSAC, 3.0);
    if (H.empty()) {
        return std::nullopt;
    }
    Homography out;
    for (int i = 0; i < 3; ++i) {
        for (int j = 0; j < 3; ++j) {
            out(i, j) = H.at<double>(i, j);
        }
    }
    return out;
}

cv::Mat warp_to_projector(const cv::Mat& cam_img, const Homography& H, const Config& cfg) {
    const cv::Matx33d inv = H.inv();
    cv::Mat warped;
    cv::warpPerspective(cam_img, warped, cv::Mat(inv), cv::Size(cfg.proj_w, cfg.proj_h));
    return warped;
}

std::optional<Homography>
calibrate_geometry(ICamera& cam, IProjector& proj, const Config& cfg) {
    const int sq = std::min(cfg.proj_w / (cfg.board_cols + 1),
                            cfg.proj_h / (cfg.board_rows + 1));
    const int bw = (cfg.board_cols + 1) * sq;
    const int bh = (cfg.board_rows + 1) * sq;
    const int ox = (cfg.proj_w - bw) / 2;
    const int oy = (cfg.proj_h - bh) / 2;

    const std::vector<cv::Point2f> proj_corners = proj_corner_table(cfg, sq, ox, oy);

    // Generate the board image with the version-appropriate ChArUco API.
    cv::Mat board_img;
#if PROCVD_ARUCO_NEW
    const cv::aruco::Dictionary dict =
        cv::aruco::getPredefinedDictionary(cv::aruco::DICT_4X4_50);
    const cv::aruco::CharucoBoard board(
        cv::Size(cfg.board_cols + 1, cfg.board_rows + 1),
        static_cast<float>(sq), static_cast<float>(sq) * 0.75f, dict);
    const cv::aruco::CharucoDetector detector(board);
    board.generateImage(cv::Size(bw, bh), board_img, /*marginSize=*/0, /*borderBits=*/1);
#else
    const cv::Ptr<cv::aruco::Dictionary> dict =
        cv::aruco::getPredefinedDictionary(cv::aruco::DICT_4X4_50);
    const cv::Ptr<cv::aruco::CharucoBoard> board = cv::aruco::CharucoBoard::create(
        cfg.board_cols + 1, cfg.board_rows + 1,
        static_cast<float>(sq), static_cast<float>(sq) * 0.75f, dict);
    board->draw(cv::Size(bw, bh), board_img, /*marginSize=*/0, /*borderBits=*/1);
#endif

    cv::Mat canvas = cv::Mat::zeros(cfg.proj_h, cfg.proj_w, CV_8UC1);
    board_img.copyTo(canvas(cv::Rect(ox, oy, bw, bh)));
    cv::Mat canvas_bgr;
    cv::cvtColor(canvas, canvas_bgr, cv::COLOR_GRAY2BGR);
    proj.show(canvas_bgr);

    for (int attempt = 0; attempt < 5; ++attempt) {
        cv::Mat frame = cam.fresh_frame();
        cv::Mat gray;
        cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);

        std::vector<cv::Point2f> charuco_corners;
        std::vector<int> charuco_ids;
#if PROCVD_ARUCO_NEW
        detector.detectBoard(gray, charuco_corners, charuco_ids);
#else
        std::vector<int> marker_ids;
        std::vector<std::vector<cv::Point2f>> marker_corners;
        cv::aruco::detectMarkers(gray, dict, marker_corners, marker_ids);
        if (!marker_ids.empty()) {
            cv::aruco::interpolateCornersCharuco(marker_corners, marker_ids, gray, board,
                                                 charuco_corners, charuco_ids);
        }
#endif
        if (charuco_ids.size() < static_cast<size_t>(cfg.min_charuco_corners)) {
            continue;
        }

        // Match by corner id — orientation is unambiguous.
        std::vector<cv::Point2f> matched_proj;
        std::vector<cv::Point2f> matched_cam;
        for (size_t i = 0; i < charuco_ids.size(); ++i) {
            const int id = charuco_ids[i];
            if (id >= 0 && id < static_cast<int>(proj_corners.size())) {
                matched_proj.push_back(proj_corners[static_cast<size_t>(id)]);
                matched_cam.push_back(charuco_corners[i]);
            }
        }
        proj.black();
        return homography_from_correspondences(matched_proj, matched_cam);
    }

    proj.black();
    return std::nullopt;
}

}  // namespace procvd

File: python/calibration/geometry.py
"""
Geometry calibration — runs on a keypress.

Projects a ChArUco board, detects it in the camera, and computes a single
homography mapping projector pixels -> camera pixels for the current surface
pose. ChArUco corner IDs make the board orientation unambiguous.

This is the ONLY geometry method. The old Gray-code / structured-light package
and the radiometric solver were removed when the pipeline was simplified to
"project diagonal stripes onto red regions".

Output (homography H) feeds the runtime: a camera-space pattern overlay is
warped into projector space with inv(H) before projection.
"""

import sys
import time
from pathlib import Path

import cv2
import numpy as np

# Ensure 'python/' is on the path whether this file is imported or run directly
_python_root = Path(__file__).parent.parent
if str(_python_root) not in sys.path:
    sys.path.insert(0, str(_python_root))

from calibration.projector import ProjectorOutput, PROJ_W, PROJ_H

# ChArUco board geometry — inner corners across/down.
BOARD_COLS = 7
BOARD_ROWS = 7
# Square size that fits the board to the projector canvas
PROJ_SQUARE_PX = min(PROJ_W // (BOARD_COLS + 1), PROJ_H // (BOARD_ROWS + 1))

SETTLE_MS = 150
MAX_TRIES = 5

# ---------------------------------------------------------------------------
# ChArUco board (chessboard + ArUco markers — unambiguous orientation)
# ---------------------------------------------------------------------------

_ARUCO_DICT = cv2.aruco.getPredefinedDictionary(cv2.aruco.DICT_4X4_50)
_CHARUCO_BOARD = cv2.aruco.CharucoBoard(
    (BOARD_COLS + 1, BOARD_ROWS + 1),   # squares across × down (outer squares)
    PROJ_SQUARE_PX,                      # square side length (projector pixels)
    PROJ_SQUARE_PX * 0.75,               # marker side length
    _ARUCO_DICT,
)
_CHARUCO_DETECTOR = cv2.aruco.CharucoDetector(_CHARUCO_BOARD)


def make_projected_charuco() -> np.ndarray:
    """Generate a ChArUco board image sized for the projector, centred."""
    sq = PROJ_SQUARE_PX
    bw = (BOARD_COLS + 1) * sq
    bh = (BOARD_ROWS + 1) * sq
    ox = (PROJ_W - bw) // 2
    oy = (PROJ_H - bh) // 2
    board_img = _CHARUCO_BOARD.generateImage((bw, bh))
    canvas = np.zeros((PROJ_H, PROJ_W), dtype=np.uint8)
    canvas[oy:oy + bh, ox:ox + bw] = board_img
    return cv2.cvtColor(canvas, cv2.COLOR_GRAY2BGR)


def _proj_charuco_corners() -> dict[int, np.ndarray]:
    """
    Return {corner_id: (x, y)} for all inner ChArUco corners in projector pixels.
    Corner IDs match the ArUco convention (row-major, 0-indexed).
    """
    sq = PROJ_SQUARE_PX
    ox = (PROJ_W - (BOARD_COLS + 1) * sq) // 2
    oy = (PROJ_H - (BOARD_ROWS + 1) * sq) // 2
    corners = {}
    cid = 0
    for r in range(BOARD_ROWS):
        for c in range(BOARD_COLS):
            corners[cid] = np.array([(ox + (c + 1) * sq), (oy + (r + 1) * sq)],
                                    dtype=np.float32)
            cid += 1
    return corners


def fresh_frame(cap: cv2.VideoCapture) -> np.ndarray:
    """Read a fresh frame, flushing a couple of buffered stale ones first."""
    for _ in range(2):
        cap.read()
    ret, frame = cap.read()
    if not ret:
        raise RuntimeError("Camera read failed during geometry calibration.")
    return frame


# ---------------------------------------------------------------------------
# Geometry — homography from projected ChArUco board
# ---------------------------------------------------------------------------

def calibrate_geometry(
    cap: cv2.VideoCapture,
    proj: ProjectorOutput,
) -> np.ndarray:
    """
    Project a ChArUco board, detect it with ArUco markers, compute homography.
    Corner IDs make orientation unambiguous.

    Returns H (3×3 float64) mapping projector pixels -> camera pixels.
    Leaves the projector black on return.
    """
    pattern = make_projected_charuco()
    proj_map = _proj_charuco_corners()   # id → projector (x, y)

    proj.show(pattern, flush=True)
    time.sleep(SETTLE_MS / 1000)

    charuco_corners = charuco_ids = None
    for _ in range(MAX_TRIES):
        frame = fresh_frame(cap)
        grey = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
        charuco_corners, charuco_ids, _, _ = _CHARUCO_DETECTOR.detectBoard(grey)
        if charuco_ids is not None and len(charuco_ids) >= 4:
            break
        time.sleep(0.05)
    else:
        proj.black()
        raise RuntimeError(
            f"Geometry calibration failed — detected "
            f"{len(charuco_ids) if charuco_ids is not None else 0} ChArUco corners "
            f"(need ≥ 4). Check lighting and that the projected board is fully visible."
        )

    proj.black()

    # Build matched point lists using corner IDs — orientation is unambiguous
    proj_matched = []
    cam_matched = []
    for i, cid in enumerate(charuco_ids.flatten()):
        if cid in proj_map:
            proj_matched.append(proj_map[cid])
            cam_matched.append(charuco_corners[i][0])

    if len(proj_matched) < 4:
        raise RuntimeError(f"Only {len(proj_matched)} matched corners — need ≥ 4.")

    proj_pts_arr = np.array(proj_matched, dtype=np.float32).reshape(-1, 1, 2)
    cam_pts_arr = np.array(cam_matched, dtype=np.float32).reshape(-1, 1, 2)

    H, mask = cv2.findHomography(proj_pts_arr, cam_pts_arr,
                                 cv2.RANSAC, ransacReprojThreshold=3.0)
    if H is None:
        raise RuntimeError("findHomography failed — not enough valid correspondences.")

    inliers = int(mask.sum()) if mask is not None else 0
    print(f"  Geometry: {inliers}/{len(proj_matched)} inliers  "
          f"({len(charuco_ids)} ChArUco corners detected)")
    return H


# ---------------------------------------------------------------------------
# Self-test (no projector needed — uses synthetic data)
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    # Verify the homography round-trip with synthetic correspondences.
    proj_pts = np.array(list(_proj_charuco_corners().values()), dtype=np.float32)

    # Synthetic camera points: apply a known homography
    H_true = np.array([[0.9, 0.05, 50],
                       [0.02, 0.88, 30],
                       [0.0001, 0.0001, 1.0]], dtype=np.float64)

    cam_pts_h = (H_true @ np.column_stack([proj_pts, np.ones(len(proj_pts))]).T).T
    cam_pts = (cam_pts_h[:, :2] / cam_pts_h[:, 2:3]).astype(np.float32)

    H_est, _ = cv2.findHomography(
        proj_pts.astype(np.float32).reshape(-1, 1, 2),
        cam_pts.reshape(-1, 1, 2),
        cv2.RANSAC,
    )

    err = np.abs(H_est / H_est[2, 2] - H_true / H_true[2, 2]).max()
    print(f"Homography recovery error: {err:.2e}")
    assert err < 1e-4, f"Homography error too high: {err}"
    print("PASS")
